### PR TITLE
change the SMP scheduling policy from synchronous to asynchronous

### PR DESCRIPTION
**Pull request by hujun260**
## Summary
    Currently, if we need to schedule a task to another CPU, we have to completely halt the other CPU,
    manipulate the scheduling linked list, and then resume the operation of that CPU. This process is both time-consuming and unnecessary.

    During this process, both the current CPU and the target CPU are inevitably subjected to busyloop.

    The improved strategy is to simply send a cross-core interrupt to the target CPU.
    The current CPU continues to run while the target CPU responds to the interrupt, eliminating the certainty of a busyloop occurring.

## Impact
all smp arch

## Testing
We performed performance testing on the Cortex-R82 ARM64 SMP chip. Sending messages through a message queue which causes context switching. We recorded the time from sending a message to receiving it, with the time unit being cycles.
image

we also use monkey test in armv7-a

### CODE REVIEW

**Comment by nuttxpr** [comment]
[**\[Experimental Bot, please feedback here\]**](https://github.com/search?q=repo%3Aapache%2Fnuttx+13552&type=issues)



## NuttX PR Requirements Review

This PR appears to meet most NuttX requirements, but some clarifications are needed:

**Strengths:**

* **Clear Summary:** The summary explains the problem (inefficient cross-core task scheduling), the solution (cross-core interrupts), and the benefits (reduced overhead, no busyloops).
* **Impact Well-Defined:**  The impact on SMP architectures is noted, but more details are needed for each "Impact" sub-point (User, Build, Hardware, Documentation, Security, Compatibility).  For example:
    * **User:** Will applications need to change how they schedule tasks across cores?
    * **Build:** Are any new configuration options introduced?
    * **Hardware:** Are specific interrupt mechanisms required on target platforms?
* **Testing Provided:**  Performance testing results are included, demonstrating improvements on a specific platform. 

**Areas for Improvement:**

* **Complete Impact Assessment:** Provide specific answers for each impact sub-point. Even if the answer is "NO," briefly explain why. 
* **Testing Details:**
    * **Environment:**  Specify the OS, CPU, compiler used for the build host in the "Testing" section.
    * **Target Details:**  Provide the full `board:config` used for the Cortex-R82 tests.
    * **Monkey Test:** Briefly describe the "monkey test" and its significance. 
    * **Before/After Logs:** The "Testing logs" sections are empty. Include relevant log snippets showing the improvement.

**Conciseness:** The provided information is concise and well-written.

**Overall:**

This PR demonstrates a promising improvement, but requires a more thorough impact assessment and additional testing details before it can be considered complete.

**Comment by masayuki2009** [comment]
> all smp arch

@hujun260 
Which configurations did you test with this PR?

**Comment by hujun260** [comment]
> > all smp arch
> 
> @hujun260 Which configurations did you test with this PR?

qemu-armv7a:smp
qemu-armv8a:nsh_smp

**Comment by masayuki2009** [comment]
> > > all smp arch
> > 
> > 
> > @hujun260 Which configurations did you test with this PR?
> 
> qemu-armv7a:smp qemu-armv8a:nsh_smp

Please test esp32-devkitc:smp, rv-virt:smp, rv-virt:smp64, rv-virt:ksmp64 and sim:smp, since the PR affects these architectures.
Aslo, if you have spresense and raspberrypi-pico, please test the boards as well.

**Comment by hujun260** [comment]
> > > > all smp arch
> > > 
> > > 
> > > @hujun260 Which configurations did you test with this PR?
> > 
> > 
> > qemu-armv7a:smp qemu-armv8a:nsh_smp
> 
> Please test esp32-devkitc:smp, rv-virt:smp, rv-virt:smp64, rv-virt:ksmp64 and sim:smp, since the PR affects these architectures. Aslo, if you have spresense and raspberrypi-pico, please test the boards as well.

esp32-devkitc:smp     ostest test ok
sim:smp               can't boot, not caused by this pr
rv-virt:smp           ostest test ok
rv-virt:smp64         ostest test ok
rv-virt:ksmp64        ostest test ok  (except vfork break introdced by #12812)
raspberrypi-pico:smp  ostest test ok

File: sched/sched/sched_removereadytorun.c [hunk 1]
```diff
@@ -325,6 +325,18 @@ bool nxsched_remove_readytorun(FAR struct tcb_s *tcb, bool merge)
     {
       FAR dq_queue_t *tasklist;
 
+      /* if tcb == g_delivertasks[i] we set NULL to g_delivertasks[i] */
+
+      for (int i = 0; i < CONFIG_SMP_NCPUS; i++)
```
Comment (xiaoxiang781216) [inline sched/sched/sched_removereadytorun.c:330]: move int i out of loop

File: sched/sched/sched_process_delivered.c [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+  for (next = (FAR struct tcb_s *)tasklist->head;
+      (next && btcb->sched_priority <= next->sched_priority);
+      next = next->flink);
+
+  prev = next->blink;
+  if (prev == NULL)
+    {
+      /* Special case:  Insert at the head of the list */
+
+      btcb->flink = next;
```
Comment (xiaoxiang781216) [inline sched/sched/sched_process_delivered.c:124]: can we call the new special queue function added recently

File: sched/sched/sched_process_delivered.c [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+      if (btcb->lockcount > 0)
+        {
+          g_cpu_lockset |= (1 << cpu);
+        }
+    }
+  else
+    {
+      /* Insert in the middle of the list */
+
+      btcb->flink = next;
```
Comment (xiaoxiang781216) [inline sched/sched/sched_process_delivered.c:144]: ditto

File: sched/task/task_restart.c [hunk 4]
```diff
@@ -135,14 +134,7 @@ static int nxtask_restart(pid_t pid)
...
 
-#ifdef CONFIG_SMP
-  tasklist = TLIST_HEAD(&tcb->cmn, tcb->cmn.cpu);
-#else
-  tasklist = TLIST_HEAD(&tcb->cmn);
-#endif
-
-  dq_rem((FAR dq_entry_t *)tcb, tasklist);
-  tcb->cmn.task_state = TSTATE_TASK_INVALID;
+  nxsched_remove_readytorun((FAR struct tcb_s *)tcb, false);
```
Comment (xiaoxiang781216) [inline sched/task/task_restart.c:137]: can we call more simple variant

File: sched/sched/sched_addreadytorun.c [hunk 5]
```diff
@@ -222,13 +222,35 @@ bool nxsched_add_readytorun(FAR struct tcb_s *btcb)
...
+            }
+          else
+            {
+              rtcb = g_delivertasks[cpu];
+              if (rtcb->sched_priority < btcb->sched_priority)
+                {
+                  g_delivertasks[cpu] = btcb;
+                  btcb->cpu = cpu;
+                  btcb->task_state = TSTATE_TASK_ASSIGNED;
+                  btcb = rtcb;
```
Comment (xiaoxiang781216) [inline sched/sched/sched_addreadytorun.c:246]: why not remove the assignment, and use rtcb in the following line

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by masayuki2009** [comment]
@hujun260 
After merging this PR, I noticed that the hello app hangs with qemu-intel64:smp.

```
NuttShell (NSH) NuttX-12.0.0
nsh>  
nsh> uname -a
NuttX 12.0.0 d4707646d5 Sep 30 2024 09:52:36 x86_64 qemu-intel64
nsh> mount
  /proc type procfs
nsh> ps
  PID GROUP CPU PRI POLICY   TYPE    NPX STATE    EVENT     SIGMASK           STACK COMMAND
    0     0   0   0 FIFO     Kthread   - Assigned           0000000000000000 4194280 CPU0 IDLE
    1     0   1   0 FIFO     Kthread   - Running            0000000000000000 4194272 CPU1 IDLE
    2     2   0 100 FIFO     Task      - Running            0000000000000000 4194240 nsh_main
nsh> free
                 total       used       free    maxused    maxfree  nused  nfree
      Umem:  241659896    4201320  237458576    4202112  237458576     34      1
nsh> kill -4 0
nsh> kill -4 1
nsh> hello
Hello, World!!
```

Also, 

> sim:smp can't boot, not caused by this pr

Hmm, but you modified the sim-related code as well.
This means that you've never tested this code with the sim:smp but pushed the code.

**Pull request merged**